src/rest-server/src/middlewares/v2/protocol.js

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Microsoft Corporation
+// All rights reserved.
+//
+// MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+
+// module dependencies
+const yaml = require('js-yaml');
+const mustache = require('mustache');
+const createError = require('../../util/error');
+const protocolSchema = require('../../config/v2/protocol');
+
+
+const prerequisiteTypes = [
+  'script',
+  'output',
+  'data',
+  'dockerimage',
+];
+
+const prerequisiteFields = [
+  'script',
+  'output',
+  'data',
+  'dockerImage',
+];
+
+const protocolValidate = (protocolYAML) => {
+  const protocolObj = yaml.safeLoad(protocolYAML);
+  if (!protocolSchema.validate(protocolObj)) {
+    throw createError('Bad Request', 'InvalidProtocolError', protocolSchema.validate.errors);
+  }
+  // convert prerequisites list to dict
+  const prerequisites = {};
+  for (let type of prerequisiteTypes) {
+    prerequisites[type] = {};
+  }
+  if ('prerequisites' in protocolObj) {
+    for (let item of protocolObj.prerequisites) {
+      prerequisites[item.type][item.name] = item;
+    }
```

Comment:
Will you check the name duplication for the same type of prerequisites?

Reply:
Nope, the prerequisite with same type and same type will **overwrite** the previous one.

Reply:
Do you think is it useful to prompt users name duplicated? Maybe users sometimes forget to change the name when they create a new prerequisites by copying and modify the existing ones?

---
In reply to: [270776412](https://github.com/Microsoft/pai/pull/2260#discussion_r270776412) [](ancestors = 270776412)

Reply:
Updated, will throw bad request error if duplicate prerequisites or deployments exist.